sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.nio.entity.NStringEntity;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestClientBuilder;
+
+/**
+ * IO to read and write data on Elasticsearch.
+ *
+ * <h3>Reading from Elasticsearch</h3>
+ *
+ * <p>ElasticsearchIO source returns a bounded collection of String representing JSON document
+ * as {@code PCollection<String>}.
+ *
+ * <p>To configure the Elasticsearch source, you have to provide a connection configuration
+ * containing the HTTP address of the instances, an index name and a type. The following example
+ * illustrates options for configuring the source:
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(ElasticsearchIO.read().withConnectionConfiguration(
+ *    ElasticsearchIO.ConnectionConfiguration.create("http://host:9200", "my-index", "my-type")
+ * )
+ *
+ * }</pre>
+ *
+ * <p>The connection configuration also accepts optional configuration: {@code withUsername()} and
+ * {@code withPassword()}.
+ *
+ * <p>You can also specify a query on the {@code read()} using {@code withQuery()}.
+ *
+ * <h3>Writing to Elasticsearch</h3>
+ *
+ * <p>ElasticsearchIO supports sink to write documents (as JSON String).
+ *
+ * <p>To configure Elasticsearch sink, similar to the read, you have to provide a connection
+ * configuration. For instance:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...)
+ *    .apply(ElasticsearchIO.write().withConnectionConfiguration(
+ *       ElasticsearchIO.ConnectionConfiguration.create("http://host:9200", "my-index", "my-type")
+ *    )
+ *
+ * }</pre>
+ *
+ * <p>Optionally, you can provide {@code withBatchSize()} and {@code withBatchSizeMegaBytes()}
+ * to specify the size of the write batch.
+ */
+public class ElasticsearchIO {
+
+  public static Read read() {
+    return new AutoValue_ElasticsearchIO_Read.Builder().build();
+  }
+
+  public static Write write() {
+    return new AutoValue_ElasticsearchIO_Write.Builder().setBatchSize(1000L)
+        .setBatchSizeMegaBytes(5).build();
+  }
+
+  private ElasticsearchIO() {
+  }
+
+  private static JsonObject parseResponse(Response response) throws IOException {
+    InputStream content = response.getEntity().getContent();
+    InputStreamReader inputStreamReader = new InputStreamReader(content, "UTF-8");
+    JsonObject jsonObject = new Gson().fromJson(inputStreamReader, JsonObject.class);
+    return jsonObject;
+  }
+
+  /**
+   * A POJO describing a connection configuration to Elasticsearch.
+   */
+  @AutoValue
+  public abstract static class ConnectionConfiguration implements Serializable {
+
+    abstract List<String> getAddresses();
+    @Nullable abstract String getUsername();
+    @Nullable abstract String getPassword();
+    abstract String getIndex();
```

Comment:
Just to confirm: all of these fields make sense both for read() and write(), correct?